File: JStream/Writer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace JStream;

public class Writer
{
    private readonly Action<string> _output;
    private readonly List<int> _stack = new();
    private bool _enableIndent = true;
    private bool _enableNewline = true;
    private bool _allowNaN = false;

    public Writer(Action<string>? output = null)
    {
        _output = output ?? Console.Write;
        _stack.Add(0);
    }

    public bool EnableIndent
    {
        get => _enableIndent;
        set => _enableIndent = value;
    }

    public bool EnableNewline
    {
        get => _enableNewline;
        set => _enableNewline = value;
    }

    public bool AllowNaN
    {
        get => _allowNaN;
        set => _allowNaN = value;
    }

    private void Print(string text) => _output(text);
    private void Print(char ch) => _output(ch.ToString());

    private void PrintNewline()
    {
        if (!_enableNewline) return;
        Print('\n');
    }

    private void PrintIndent()
    {
        if (!_enableIndent) return;
        
        int depth = _stack.Count - 1;
        for (int i = 0; i < depth; i++)
        {
            Print("  ");
        }
    }

    private bool PrintNumber(double value)
    {
        string formatted = JsonHelper.FormatNumber(value, _allowNaN);
        if (formatted == "null" && !_allowNaN)
        {
            Print("null");
            return false;
        }
        
        Print(formatted);
        return true;
    }

    private void PrintString(string value)
    {
        Print('"');
        Print(JsonHelper.EncodeJsonString(value));
        Print('"');
    }

    private bool PrintValue(string name, Func<bool> fn)
    {
        PrintName(name);
        bool result = fn();
        
        if (_stack.Count > 0)
            _stack[^1]++;
        
        return result;
    }

    private void PrintObject(string name = "", Action? fn = null)
    {
        PrintName(name);
        Print('{');
        _stack.Add(0);
        
        if (fn != null)
        {
            fn();
            EndObject();
        }
    }

    private void PrintArray(string name = "", Action? fn = null)
    {
        PrintName(name);
        Print('[');
        _stack.Add(0);
        
        if (fn != null)
        {
            fn();
            EndArray();
        }
    }

    private void EndBlock()
    {
        PrintNewline();
        
        if (_stack.Count > 0)
        {
            _stack.RemoveAt(_stack.Count - 1);
            if (_stack.Count > 0)
                _stack[^1]++;
        }
        
        PrintIndent();
    }

    public void PrintName(string name)
    {
        if (_stack.Count > 0 && _stack[^1] > 0)
            Print(',');
        
        if (_stack.Count > 1)
            PrintNewline();
        
        PrintIndent();
        
        if (!string.IsNullOrEmpty(name))
        {
            PrintString(name);
            Print(':');
            if (_enableIndent)
                Print(' ');
        }
    }

    public void StartObject(string name = "")
    {
        PrintObject(name);
    }

    public void EndObject()
    {
        EndBlock();
        Print('}');
    }

    public void Object(string name, Action fn)
    {
        PrintObject(name, fn);
    }

    public void StartArray(string name = "")
    {
        PrintArray(name);
    }

    public void EndArray()
    {
        EndBlock();
        Print(']');
    }

    public void Array(string name, Action fn)
    {
        PrintArray(name, fn);
    }

    public bool Number(string name, double value)
    {
        return PrintValue(name, () => PrintNumber(value));
    }

    public void Number(double value)
    {
        Number("", value);
    }

    public void String(string name, string value)
    {
        PrintValue(name, () =>
        {
            PrintString(value);
            return true;
        });
    }

    public void String(string value)
    {
        String("", value);
    }

    public void Symbol(string name, StateType value)
    {
        PrintValue(name, () =>
        {
            string text = value switch
            {
                StateType.False => "false",
                StateType.True => "true",
                _ => "null"
            };
            Print(text);
            return true;
        });
    }

    public void Boolean(string name, bool value)
    {
        Symbol(name, value ? StateType.True : StateType.False);
    }

    public void Boolean(bool value)
    {
        Boolean("", value);
    }

    public void Null(string name = "")
    {
        Symbol(name, StateType.Null);
    }

    public void WriteVariant(string name, Variant variant)
    {
        if (variant.IsNull)
        {
            Null(name);
        }
        else if (variant.IsBoolean)
        {
            Boolean(name, variant.Get<bool>());
        }
        else if (variant.IsNumber)
        {
            Number(name, variant.Get<double>());
        }
        else if (variant.IsString)
        {
            String(name, variant.Get<string>());
        }
        else if (variant.IsObject)
        {
            var obj = variant.Get<JObject>();
            Object(name, () =>
            {
                foreach (var kvp in obj)
                {
                    WriteVariant(kvp.Key, kvp.Value);
                }
            });
        }
        else if (variant.IsArray)
        {
            var array = variant.Get<JArray>();
            Array(name, () =>
            {
                foreach (var item in array)
                {
                    WriteVariant("", item);
                }
            });
        }
    }

    public void WriteVariant(Variant variant)
    {
        WriteVariant("", variant);
    }

    public void Finish()
    {
        if (_stack.Count > 0 && _stack[0] > 0)
            PrintNewline();
    }
}

File: jstream-go/writer.go
package jstream

import "strings"

// Writer is a streaming JSON generator.
type Writer struct {
	output        func(string)
	sb            strings.Builder
	stack         []int
	enableIndent  bool
	enableNewline bool
	allowNaN      bool
}

// NewWriter creates a Writer that collects output in a string.
func NewWriter() *Writer {
	return &Writer{
		stack:         []int{0},
		enableIndent:  true,
		enableNewline: true,
	}
}

// NewWriterWithOutput creates a Writer that writes chunks to the given callback.
func NewWriterWithOutput(output func(string)) *Writer {
	w := NewWriter()
	w.output = output
	return w
}

func (w *Writer) EnableIndent(enabled bool)  { w.enableIndent = enabled }
func (w *Writer) EnableNewline(enabled bool) { w.enableNewline = enabled }
func (w *Writer) AllowNaN(allow bool)        { w.allowNaN = allow }

func (w *Writer) print(s string) {
	if w.output != nil {
		w.output(s)
	} else {
		w.sb.WriteString(s)
	}
}

func (w *Writer) printNewline() {
	if !w.enableNewline {
		return
	}
	w.print("\n")
}

func (w *Writer) printIndent() {
	if !w.enableIndent {
		return
	}
	depth := len(w.stack) - 1
	for i := 0; i < depth; i++ {
		w.print("  ")
	}
}

func (w *Writer) printString(s string) {
	w.print("\"" + encodeJSONString(s) + "\"")
}

func (w *Writer) printNumber(v float64) bool {
	s := formatNumber(v, w.allowNaN)
	if s == "null" && !w.allowNaN {
		w.print("null")
		return false
	}
	w.print(s)
	return true
}

func (w *Writer) printValue(name string, fn func() bool) bool {
	w.printName(name)
	ok := fn()
	if len(w.stack) > 0 {
		w.stack[len(w.stack)-1]++
	}
	if len(w.stack) == 1 {
		w.flush()
	}
	return ok
}

// WriteName writes a property name (including comma and indentation).
func (w *Writer) WriteName(name string) {
	w.printName(name)
}

func (w *Writer) printName(name string) {
	if len(w.stack) > 0 && w.stack[len(w.stack)-1] > 0 {
		w.print(",")
	}
	if len(w.stack) > 1 {
		w.printNewline()
	}
	w.printIndent()
	if name != "" {
		w.printString(name)
		w.print(":")
		if w.enableIndent {
			w.print(" ")
		}
	}
}

func (w *Writer) printObject(name string, fn func()) {
	w.printName(name)
	w.print("{")
	w.stack = append(w.stack, 0)
	if fn != nil {
		fn()
		w.EndObject()
	}
}

func (w *Writer) printArray(name string, fn func()) {
	w.printName(name)
	w.print("[")
	w.stack = append(w.stack, 0)
	if fn != nil {
		fn()
		w.EndArray()
	}
}

func (w *Writer) endBlock() {
	w.printNewline()
	if len(w.stack) > 0 {
		w.stack = w.stack[:len(w.stack)-1]
		if len(w.stack) > 0 {
			w.stack[len(w.stack)-1]++
		}
	}
	w.printIndent()
}

func (w *Writer) flush() {
	if len(w.stack) > 0 && w.stack[0] > 0 {
		w.printNewline()
	}
	w.stack = []int{0}
}

// StartObject begins a new object.
func (w *Writer) StartObject(name ...string) {
	n := ""
	if len(name) > 0 {
		n = name[0]
	}
	w.printObject(n, nil)
}

// EndObject ends the current object.
func (w *Writer) EndObject() {
	w.endBlock()
	w.print("}")
	if len(w.stack) == 1 {
		w.flush()
	}
}

// Object writes a named object using the provided callback.
func (w *Writer) Object(name string, fn func()) {
	w.printObject(name, fn)
}

// StartArray begins a new array.
func (w *Writer) StartArray(name ...string) {
	n := ""
	if len(name) > 0 {
		n = name[0]
	}
	w.printArray(n, nil)
}

// EndArray ends the current array.
func (w *Writer) EndArray() {
	w.endBlock()
	w.print("]")
	if len(w.stack) == 1 {
		w.flush()
	}
}

// Array writes a named array using the provided callback.
func (w *Writer) Array(name string, fn func()) {
	w.printArray(name, fn)
}

// WriteNumber writes a number value.
func (w *Writer) WriteNumber(name string, value float64) bool {
	return w.printValue(name, func() bool { return w.printNumber(value) })
}

// WriteNumberValue writes an unnamed number value.
func (w *Writer) WriteNumberValue(value float64) bool {
	return w.WriteNumber("", value)
}

// WriteString writes a string value.
func (w *Writer) WriteString(name string, value string) {
	w.printValue(name, func() bool {
		w.printString(value)
		return true
	})
}

// WriteStringValue writes an unnamed string value.
func (w *Writer) WriteStringValue(value string) {
	w.WriteString("", value)
}

// WriteSymbol writes a symbolic value (false, true, or null).
func (w *Writer) WriteSymbol(name string, value StateType) {
	w.printValue(name, func() bool {
		switch value {
		case StateFalse:
			w.print("false")
		case StateTrue:
			w.print("true")
		default:
			w.print("null")
		}
		return true
	})
}

// WriteBoolean writes a boolean value.
func (w *Writer) WriteBoolean(name string, value bool) {
	if value {
		w.WriteSymbol(name, StateTrue)
	} else {
		w.WriteSymbol(name, StateFalse)
	}
}

// WriteBooleanValue writes an unnamed boolean value.
func (w *Writer) WriteBooleanValue(value bool) {
	w.WriteBoolean("", value)
}

// WriteNull writes a null value.
func (w *Writer) WriteNull(name ...string) {
	n := ""
	if len(name) > 0 {
		n = name[0]
	}
	w.WriteSymbol(n, StateNull)
}

// WriteRaw writes raw JSON text. The caller is responsible for its validity.
func (w *Writer) WriteRaw(name string, value string) {
	w.printValue(name, func() bool {
		w.print(value)
		return true
	})
}

// WriteRawValue writes unnamed raw JSON text.
func (w *Writer) WriteRawValue(value string) {
	w.WriteRaw("", value)
}

// Finish writes a final newline if content was produced.
func (w *Writer) Finish() {
	if len(w.stack) > 0 && w.stack[0] > 0 {
		w.printNewline()
	}
}

// String returns the collected output.
func (w *Writer) String() string {
	return w.sb.String()
}
